autoflip_calculators/shot_change_fusing/shot_change_fusing_calculator.cc

```diff
@@ -0,0 +1,273 @@
+  ShotChangeFusingCalculatorOptions options_;
+  // Key is the id and value is the priority
+  std::map<int32, int32> priority_;
+  std::vector<ShotSignal> shot_signals_;
+  bool tag_input_interface_;
+//   bool process_by_scene_;
+};
+
+REGISTER_CALCULATOR(ShotChangeFusingCalculator);
+
+namespace {
+void SetupTagInput(mediapipe::CalculatorContract* cc) {
+  for (int i = 0; i < cc->Inputs().NumEntries(kIsShotBoundaryTag); ++i) {
+    cc->Inputs().Get(kIsShotBoundaryTag, i).Set<bool>();
+  }
+  cc->Outputs().Tag(kOutputTag).Set<bool>();
+}
+
+void SetupOrderedInput(mediapipe::CalculatorContract* cc) {
+  for (int i = 0; i < cc->Inputs().NumEntries(); ++i) {
+    cc->Inputs().Index(i).Set<bool>();
+  }
+  cc->Outputs().Index(0).Set<bool>();
+}
+}  // namespace
+
+::mediapipe::Status ShotChangeFusingCalculator::GetContract(
+    mediapipe::CalculatorContract* cc) {
+  if (cc->Inputs().NumEntries(kIsShotBoundaryTag) > 0) {
+    SetupTagInput(cc);
+  } else {
+    SetupOrderedInput(cc);
+  }
+  return ::mediapipe::OkStatus();
+}
+
+mediapipe::Status ShotChangeFusingCalculator::Open(
+    mediapipe::CalculatorContext* cc) {
+  options_ = cc->Options<ShotChangeFusingCalculatorOptions>();
+  RET_CHECK_EQ(options_.shot_settings().size(), cc->Inputs().NumEntries())
+    << "The number of signals does not equal to the number of signal settings.";
+
+  if (cc->Inputs().HasTag(kIsShotBoundaryTag)) {
+    tag_input_interface_ = true;
+  }
+  for (int count = 0; count < options_.shot_settings().size(); ++count) {
+    const auto& setting = options_.shot_settings()[count];
+    if (tag_input_interface_) {
+      RET_CHECK(setting.id() != kIdInitial) 
+        << "Shot change id is missing in tag interface";
+      RET_CHECK(priority_.find(setting.id()) ==  priority_.end())
+        << "Duplicate shot change id: " << setting.id();
+      priority_[setting.id()] = setting.priority();
+    }
+    else {
+      priority_[count] = setting.priority();
+    }
+  }
+
+  return ::mediapipe::OkStatus();
+}
+
+mediapipe::Status ShotChangeFusingCalculator::Process(
+    mediapipe::CalculatorContext* cc) {
+  ShotSignal signal;
+  const auto& signal_packets = GetSignalPackets(cc);
+  for (int i = 0; i < signal_packets.size(); ++i) {
+    const auto& packet = signal_packets[i];
+    if (packet.IsEmpty()) {
+      continue;
+    }
+    const bool boundary = packet.Get<bool>();
+    signal.shot_change_signal = signal.shot_change_signal || boundary;
+    // Keep the highest priority.
+    signal.priority = std::min(signal.priority, priority_[i]);
+  }
+  // Store the signal only when there is input and there is a shot change.
+  if (signal.priority != kInfity || signal.shot_change_signal) {
+    signal.time = cc->InputTimestamp();
+    shot_signals_.push_back(signal);
+  }
+
+  // Flush buffer on same input if it exceeds max_scene_size or if there is not
+  // shot input information.
+  if (!shot_signals_.empty() && shot_signals_.size() > options_.max_scene_size()) {
```

Comment:
I'm not seeing a reason to do caching in this calculator.  I would just always output the signal on each input.  No need for "shot_signals_" in that case I think.

Reply:
To cache because we want to merge some signals. Since we may not detect the shot boundary and speaker very exactly. Sometimes, two signals are very close, for example, 22.022s and 22.322s. I think it is better to merge this two. Does this make sense?